### DISCUSSION TITLE
Base64 encode/decode JS in automation - atob/btoa

### DISCUSSION THREAD

**Discussion by andyburgessmd** [Budibase/budibase#13148] [Help]
I'm trying to use the atob JS function in automation to decode a base64 encoded string but the output is always 'Error while executing JS'.
The JS works when used in a page but not in automation.
For now I am just trying to get a very simple 'hello world' to decode with   
```
var decoded =  atob("aGVsbG8gd29ybGQ=");
return decoded;
```
in a backend log but this returns the error above.

Does anyone know if atob/btoa are supported in automation?

**Comment by ConorWebb96** [comment]
Hey @andyburgessmd,

TLDR of the below: atob() won't work in backend development environments.

The reason for this is due to automations being a backend piece of code. The atob() is commonly associated with frontend development, particularly in web browsers, where it is used to decode data that has been encoded using the btoa() function. However, it's important to note that the atob() and btoa() functions are part of the Web API and are not native to Node.js or other backend JavaScript environments.

We have decode helper functions that you can use.

```js
var decoded =  helpers.decodeURI("aGVsbG8gd29ybGQ=")
return decoded;
```
<img width="1252" alt="Screenshot 2024-03-01 at 08 28 25" src="https://github.com/Budibase/budibase/assets/126772285/673a2d0e-d144-47de-a56e-d2aa80995661">

**Reply by andyburgessmd** [reply]
Are there any other helpers that could do this? Even your example using decodeURI fails to decode the string to 'hello world'.

I'm trying to extract the comments added using the comment plugin from @aptkingston so the comment can be sent in an email. I just can't decode it in the automation as atob is not available there. Any other solutions you can think of?

**Reply by ConorWebb96** [reply]
Hey @andyburgessmd,

Can you decode it before you pass it onto an automation within the design area using atob?

You could try this approach.

```js
function decodeEncodedArray(encodedArray) {
    const decodedString = decodeURIComponent(encodedArray);
    const decodedArray = JSON.parse(decodedString);

    return decodedArray;
}

const encodedArray = your_array_binding;
const decodedArray = decodeEncodedArray(encodedArray);
return decodedArray;
```

If you could give me a small snippet of the encoded array I could look at it more closely for you.

**Reply by andyburgessmd** [reply]
The comment plugin stores the comments added through it into a db field like
`eyJtZXNzYWdlIjoiSXMgdGhpcyBjb21wbGV0ZWQ/IiwiZW1haWwiOiJlbWFpbEBlbWFpbC5jb20iLCJuYW1lIjoiQW5kcmV3IiwidGltZXN0YW1wIjoxNzA5MDMwMjYyMjk0fQ==|eyJtZXNzYWdlIjoiQW5vdGVociBjb21tbnQiLCJlbWFpbCI6ImVtYWlsQGVtYWlsLmNvbSIsIm5hbWUiOiJBbmRyZXciLCJ0aW1lc3RhbXAiOjE3MDkwMzAyNjIyOTR9`
Where the encoded comment objects are separated by | (pipes). 

I don't understand where you are suggesting to decode it?

**Reply by aptkingston** [reply]
Feel free to raise a feature request @andyburgessmd and we can add atob and btoa to both the frontend JS sandbox and the backend one for automations 👍 would be a pretty useful addition.

**Reply by ConorWebb96** [reply]
You can basically convert this via `atob` in the design view and then send the decoded version to an automation via a trigger automation step.

```js
return function(encodedString) {
    const parts = encodedString.split('|');
    const decodedParts = parts.map(part => atob(part));

    return decodedParts;
}
```

I've made the above into a snippet, its not needed as atob will work in the design screen regardless. This can decode your encodedString and then send it back to an automation.

**Example of the output in design.**
<img width="1196" alt="Screenshot 2024-04-19 at 16 50 40" src="https://github.com/Budibase/budibase/assets/126772285/cc0e14f0-7c37-49c0-924d-171d4aba602a">

Button setup for the trigger automation below
<img width="1199" alt="Screenshot 2024-04-19 at 16 51 26" src="https://github.com/Budibase/budibase/assets/126772285/cc9f7596-c777-4b47-aff6-0216af8b0759">

**Automation output**
<img width="388" alt="Screenshot 2024-04-19 at 16 51 51" src="https://github.com/Budibase/budibase/assets/126772285/1c773c91-802a-4c46-8be0-182bc7e3fcf8">

Then from here you can likely just loop through the array and output it to an email template.

**Reply by andyburgessmd** [reply]
Unfortunately the comments plugin doesn't allow me to add additional actions to the button.
![image](https://github.com/Budibase/budibase/assets/110813989/9136f265-f845-44d3-b5bd-1ad4b988cd41)

I did have a revelation that I could add it to a db formula field but have found that although the live evaluation view shows it works
![image](https://github.com/Budibase/budibase/assets/110813989/899ea84c-8454-4d7d-b5d6-be0ad10e775b)
when it runs the formula it returns a 'Error while executing JS' error.
![image](https://github.com/Budibase/budibase/assets/110813989/e9845d69-715e-4928-85d7-30726dc31376)
Should I raise that one as a bug?
